src/main/java/net/minecraftforge/common/block/spreader/SpreadBehaviors.java

```diff
@@ -0,0 +1,109 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.block.spreader;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.properties.BlockStateProperties;
+import net.minecraftforge.common.extensions.IForgeSpreadingBlock;
+import net.minecraftforge.registries.IRegistryDelegate;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SpreadBehaviors
+{
+
+    private static final Logger LOGGER = LogManager.getLogger();
+    private static final Map<IRegistryDelegate<Block>, Map<SpreaderType, ISpreadingBehavior>> SPREADERS = new HashMap<>();
+
+    static
+    {
+        setupVanillaBehavior();
+    }
+
+    /**
+     * This method allows you to add spreading behavior, which is used by {@link IForgeSpreadingBlock}
+     *
+     * @param block
+     * @param type
+     * @param behavior
+     */
+    public static void addSpreaderBehavior(Block block, SpreaderType type, ISpreadingBehavior behavior)
+    {
+        Map<SpreaderType, ISpreadingBehavior> blockSpecificSpreaders = SPREADERS.computeIfAbsent(block.delegate, key -> new HashMap<>());
+        if (blockSpecificSpreaders.containsKey(type))
+            LOGGER.info("Replacing spreading behavior for block '{}' and spreader type '{}'", block.getRegistryName(), type.getName());
+        blockSpecificSpreaders.put(type, behavior);
+    }
+
+    public static boolean canSpread(BlockState state, SpreaderType type)
+    {
+        return getSpreadingBehavior(state.getBlock(), type) != null;
+    }
+
+    public static BlockState getSpreadState(BlockState state, Level level, BlockPos pos, SpreaderType type)
+    {
+        ISpreadingBehavior behavior = getSpreadingBehavior(state.getBlock(), type);
+        if (behavior == null) return state;
+        return behavior.getSpreadingState(state, level, pos);
+    }
+
+    @Nullable
+    private static ISpreadingBehavior getSpreadingBehavior(Block block, SpreaderType type)
+    {
+        return SPREADERS.containsKey(block) ? SPREADERS.get(block).get(type) : null;
```

Comment:
This should be querying using the `delegate`. Also, this may benefit from doing `SPREADERS.getOrDefault(block.delegate, Collections.emptyMap()).get(type)` to avoid the dual query cost (`containsKey` then `get` if the former is successful).